teleport_tube.lua

```diff
@@ -12,47 +13,81 @@ local function hash(pos)
 end
 
 local function save_tube_db()
-	local file, err = io.open(filename, "w")
-	if file then
-		tp_tube_db.version = tp_tube_db_version
-		file:write(minetest.serialize(tp_tube_db))
-		tp_tube_db.version = nil
-		io.close(file)
-	else
-		error(err)
+	-- reset tp-tube cache
+	cache = {}
+
+	local fields = {version = tp_tube_db_version}
+	for key, val in pairs(tp_tube_db) do
+		fields[key] = minetest.serialize(val)
 	end
+	storage:from_table({fields = fields})
+end
+
+local function save_tube_db_entry(hash)
 	-- reset tp-tube cache
 	cache = {}
+
+	local val = tp_tube_db[hash]
+	storage:set_string(hash, val and minetest.serialize(val) or "")
 end
 
 local function migrate_tube_db()
+	local old_version = tp_tube_db.version or 0
+	tp_tube_db.version = nil
+	if old_version < 2.0 then
 		local tmp_db = {}
-		tp_tube_db.version = nil
 		for _, val in pairs(tp_tube_db) do
 			if(val.channel ~= "") then -- skip unconfigured tubes
 				tmp_db[hash(val)] = val
 			end
 		end
 		tp_tube_db = tmp_db
-		save_tube_db()
+	end
+	save_tube_db()
 end
 
 local function read_tube_db()
-	local file = io.open(filename, "r")
-	if file ~= nil then
+	local file = not storage:contains("version") and io.open(filename, "r")
+	if not file then
+		tp_tube_db = {}
+
+		for key, val in pairs(storage:to_table().fields) do
+			if tonumber(key) then
+				tp_tube_db[key] = minetest.deserialize(val)
+			elseif key == "version" then
+				tp_tube_db[key] = tonumber(val)
+			else
+				error("Unknown field in teleport tube DB: " .. key)
+			end
+		end
+
+		if tp_tube_db.version == nil then
+			tp_tube_db.version = tp_tube_db_version
+			storage:set_string("version", tp_tube_db.version)
+		elseif tp_tube_db.version > tp_tube_db_version then
+			error("Cannot read teleport tube DB of version " .. tp_tube_db.version)
+		end
+	else
 		local file_content = file:read("*all")
 		io.close(file)
 
+		local backup_filename = filename .. ".bak"
+		pipeworks.logger("Moving teleport tube DB to mod storage from " .. filename)
+		pipeworks.logger("Backing up old file as " .. backup_filename)
+		assert(os.rename(filename, backup_filename))
+
```

Comment:
I'd say remove this file handling stuff completely and add note to readme about it.
Could wait for other opinions on this, that's just how I feel about it: no unnecessary stuff (also will it replace existing files or should there be check for that?)

Reply:
I suppose that would work too. It's up to the maintainers.

Reply:
S-S-X makes a valid point: renaming means there would need to be one more check

Reply:
Will do then.

Reply:
Are you sure the message should be moved to the readme? IMO people are probably more likely to see it in the log.

Reply:
> Are you sure the message should be moved to the readme? IMO people are probably more likely to see it in the log.

This, keep it in the log. As almost every other mod doing stuff like this sends it to the log/print/both. Any sys admin worth there salt is going to dig through a log to see what happened first, putting it in the Readme is unitutive

Reply:
@TurkeyMcMac probably worth it to add a minetest.log line to the pipeworks.logger function, just for convenience

Reply:
Keep logging, I mean note about old db file that can be removed. Logging is good but should happen only once when automatic upgrade happens while readme should tell where data is stored and tell what's changed.

Reply:
@wsor4035 Do you mean that `pipeworks.logger` should just call `minetest.log`, or that it should call `minetest.log` in addition to `print`? I would favor the former.

Reply:
i was thinking
```lua
pipeworks.logger = function(msg)
	  print("[pipeworks]: " .. msg)
	  minetest.log("action", "[pipeworks]: " .. msg)
end
```
however if you want to drop print, i wouldnt be opposed

Reply:
I'd vote just log without print, depending on OS/config log can anyway end up to console and there print might just duplicate stuff printed to console.